### ISSUE TITLE
Loops are broken af

### ISSUE THREAD

**Issue by MrMorkat** [MrMorkat/BfInJava#1]
ArrayIndexOutOfBoundarys ¯\_(ツ)_/¯

**Comment by rdebath** [comment]
Indeed, some test cases might help; the first should say `OK` the second is a `Hello World!` both have a newline on the end.
 
``` bf
>-[+]<++++++++[>++++++++++<-]>-.>[<+>[-]]<----.[-]++++++++++.
```
``` bf
>++++++++[-<+++++++++>]<.>[][<-]>+>-[+]++>++>+++[>[->+++<<+++>]<<]>-----.
>->+++..+++.>-.<<+[>[+>+]>>]<--------------.>>.+++.------.--------.>+.>+.
```

**Comment by MrMorkat** [comment]
First one is working, second one throws an ArrayIndexOutOfBoundsException: -1 at line 79 :thinking:

**Comment by MrMorkat** [comment]
There is an empty loop in the second test case, is that one placed on purpose? The program works until it is trying to enter that loop

**Comment by rdebath** [comment]
The first one is supposed to say `OK` not `OL` so it's not working.

The second one still shows your problem with nested loops when you remove the bug checks that other interpreters trip (including the empty loop that you're not going in).
``` bf
>++++++++[-<+++++++++>]<.>      >+>    ++>++>+++[>[->+++<<+++>]<<]>-----.
>->+++..+++.>-.<<+[>[+>+]>>]<--------------.>>.+++.------.--------.>+.>+.
```

**Comment by MrMorkat** [comment]
ok, so basically most programs work as they should now, but some still go out of boundaries to the right (positive) side.

**Comment by rdebath** [comment]
Sorry, they work perfectly with an interpreter with correctly implemented while loops.
They do not need cells to the right of the starting position and they use no more that seven cells total.

This perl translator runs them all and it has no negative tape cells.
```perl
%c=qw(> $p++ < $p-- + D++ - D-- [ D&=255;while(D){ ] D&=255;} . print+chrD , D=ord(getc));
$/=$,;$_=<>;s/./$c{$&};/g;s[D]'$b[$p]'g;eval
```

And this is Urban Müller's original Hello world, which is not working now either.
``` bf
>+++++++++[<++++++++>-]<.>+++++++[<++++>-]<+.+++++++..+++.[-]>++++++++[<++++>-]
<.#>+++++++++++[<+++++>-]<.>++++++++[<+++>-]<.+++.------.--------.[-]>++++++++[
<++++>-]<+.[-]++++++++++.
```

**Comment by MrMorkat** [comment]
Well, i don't need you to tell me that they are not working, that's why i opened this issue in tge first place, i would much more appreciate it if you could tell me what exactly is wrong with my implementation :P

**Comment by rdebath** [comment]
Sigh, you said that "_basically most programs work as they should now_" which says that you think your interpreter is working correctly.

And I have already said that you are not processing nested loops correctly. 
Quoting Wikipedia:
> if the byte at the data pointer is zero, then instead of moving the instruction pointer forward to the next command, jump it forward to the command after the matching `]` command.

It says "_the matching `]` command_" key word "_matching_"; while loops nest, so when you skip over a while loop you skip over the nested ones too.

https://github.com/rdebath/Brainfuck/blob/master/bitwidth.b

**Comment by MrMorkat** [comment]
well but you're not the customer, just a random guy that decided to help me :D 
Thanks for the tip with nested loops i think i overread that.
I never said that it is working completely fine tho, otherwise i would have closed the issue.
Thanks anyways, should be easy to fix now!
